module/zfs/mmp.c

```diff
@@ -0,0 +1,316 @@
+		l = spa_get_random(VDEV_LABELS);
+		zio = zio_root(spa, NULL, NULL, flags);
+		vd->vdev_mmp_pending = gethrtime();
+		vdev_mmp_write(zio, vd, l, n, abd, mmp_write_done,
+		    mmp, flags | ZIO_FLAG_DONT_PROPAGATE);
+		zio_nowait(zio);
+	}
+	spa_config_exit(spa, SCL_STATE, FTAG);
+}
+
+static void
+mmp_thread(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+	spa_t *spa = dp->dp_spa;
+	abd_t *abd;
+	char *buf;
+	int bufsize = (1ULL << MAX_UBERBLOCK_SHIFT);
+	uint last_zfs_mmp_interval = zfs_mmp_interval;
+
+	abd = abd_alloc(bufsize, B_TRUE);
+	buf = kmem_alloc(bufsize, KM_SLEEP);
+	bzero(buf, bufsize);
+
+	mmp_thread_enter(mmp);
+
+	mmp->mmp_last_write = gethrtime();
+	dprintf("mmp_thread entered: pool %p\n", dp);
+
+	for (;;) {
+		hrtime_t start, next_time;
+
+		if (mmp->mmp_thread_exiting)
+			goto cleanup_and_exit;
+
+		start = gethrtime();
+		next_time = start + MSEC2NSEC(zfs_mmp_interval) /
+		    vdev_count_leaves(dp->dp_spa);
+
+		/*
+		 * When MMP goes off => on, no writes occurred
+		 * recently.  We update mmp_last_write to give
+		 * us some time to try.
+		 */
+		if (!last_zfs_mmp_interval && zfs_mmp_interval) {
+			dprintf("mmp_thread zfs_mmp_interval transtion: "
+			    "pool %p last_zfs_mmp_interval %lu "
+			    "zfs_mmp_interval %lu\n", dp,
+			    last_zfs_mmp_interval, zfs_mmp_interval);
+			mutex_enter(&mmp->mmp_io_lock);
+			mmp->mmp_last_write = gethrtime();
+			mutex_exit(&mmp->mmp_io_lock);
+		}
+
+		/*
+		 * Check after the transition check above.
+		 */
+		if (zfs_mmp_interval && mmp->mmp_last_write < (start -
+		    MSEC2NSEC(zfs_mmp_interval) * zfs_mmp_fail_intervals)) {
+			dprintf("mmp suspending pool: pool %s zfs_mmp_interval %lu "
+			    "start %llu mmp_last_write %llu interval_ns %llu "
+			    "zfs_mmp_fail_intervals %lu\n", spa->spa_name,
+			    zfs_mmp_interval, start, mmp->mmp_last_write, 
+			    MSEC2NSEC(zfs_mmp_interval), zfs_mmp_fail_intervals);
+			zio_suspend(spa, NULL);
+		}
+
+		mmp->mmp_last_write = gethrtime();
+		if (zfs_mmp_interval)
+			mmp_write_uberblock(dp, abd, buf, bufsize);
+
+		if (gethrtime() >= next_time) {
+			dprintf("mmp_thread skipped wait: pool %p start=%llu "
+			    "next_time=%llu now=%llu\n", dp, start, next_time,
+			    gethrtime());
+			continue;
+		}
+
+		dprintf("mmp_thread entering wait: pool %p now=%llu "
+		    "next_time=%lld\n", dp, gethrtime(), next_time);
+		(void) cv_timedwait_hires(&mmp->mmp_thread_cv,
+		    &mmp->mmp_thread_lock, next_time, NANOSEC,
+		    CALLOUT_FLAG_ABSOLUTE);
+	}
+
+cleanup_and_exit:
+	kmem_free(buf, bufsize);
+	abd_free(abd);
+
+	mmp_thread_exit(mmp, &mmp->mmp_thread);
+}
+
+#if defined(_KERNEL) && defined(HAVE_SPL)
+EXPORT_SYMBOL(mmp_init);
+EXPORT_SYMBOL(mmp_fini);
+EXPORT_SYMBOL(mmp_thread_start);
+EXPORT_SYMBOL(mmp_thread_stop);
+
+module_param(zfs_mmp_fail_intervals, uint, 0644);
+MODULE_PARM_DESC(zfs_mmp_fail_intervals, "Require MMP write in time "
+"zfs_mmp_fail_intervals*zfs_mmp_interval");
```

Comment:
It would be nice if setting `zfs_mmp_fail_intervals = 0` disabled this additional sanity check.